tower-http/src/decompression.rs

```diff
@@ -0,0 +1,692 @@
+    ResBody: Body,
+{
+    type Response = Response<DecompressBody<ResBody>>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        if !req.headers().contains_key(RANGE) {
+            if let header::Entry::Vacant(e) = req.headers_mut().entry(ACCEPT_ENCODING) {
+                if let Some(accept) = self.accept.to_header_value() {
+                    e.insert(accept);
+                }
+            }
+        }
+        ResponseFuture {
+            inner: self.inner.call(req),
+            accept: self.accept,
+        }
+    }
+}
+
+impl DecompressLayer {
+    /// Creates a new `DecompressLayer`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Sets whether to request the gzip encoding.
+    #[cfg(feature = "decompression-gzip")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "decompression-gzip")))]
+    pub fn gzip(self, enable: bool) -> Self {
+        self.accept.set_gzip(enable);
+        self
+    }
+
+    /// Sets whether to request the Deflate encoding.
+    #[cfg(feature = "decompression-deflate")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "decompression-deflate")))]
+    pub fn deflate(self, enable: bool) -> Self {
+        self.accept.set_deflate(enable);
+        self
+    }
+
+    /// Sets whether to request the Brotli encoding.
+    #[cfg(feature = "decompression-br")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "decompression-br")))]
+    pub fn br(self, enable: bool) -> Self {
+        self.accept.set_br(enable);
+        self
+    }
+
+    /// Disables the gzip encoding.
+    ///
+    /// This method is available even if the `gzip` crate feature is disabled.
+    pub fn no_gzip(self) -> Self {
+        self.accept.set_gzip(false);
+        self
+    }
+
+    /// Disables the Deflate encoding.
+    ///
+    /// This method is available even if the `deflate` crate feature is disabled.
+    pub fn no_deflate(self) -> Self {
+        self.accept.set_deflate(false);
+        self
+    }
+
+    /// Disables the Brotli encoding.
+    ///
+    /// This method is available even if the `br` crate feature is disabled.
+    pub fn no_br(self) -> Self {
+        self.accept.set_br(false);
+        self
+    }
+}
+
+impl<S> Layer<S> for DecompressLayer {
+    type Service = Decompress<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        Decompress {
+            inner: service,
+            accept: self.accept,
+        }
+    }
+}
+
+impl<F, B> Future for ResponseFuture<F>
+where
+    F: TryFuture<Ok = Response<B>>,
+    B: Body,
+{
+    type Output = Result<Response<DecompressBody<B>>, F::Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let res = ready!(self.as_mut().project().inner.try_poll(cx)?);
```

Comment:
```suggestion
        let res = ready!(self.as_mut().project().inner.poll(cx)?);
```

I think this should work if using `Future` as the trait bound.